### DISCUSSION TITLE
joinedload leads to memory leaks in combination with fastapi

### DISCUSSION THREAD

**Discussion by makatz** [sqlalchemy/sqlalchemy#11739] [Usage Questions]
Hey there and thanks for your work!

**Preamble:**
I have a few questions in regards to the use of joinedload as option in a query. We use it to populate related objects that we might use in returns of our API (FastAPI) and experience memory leaks. 

The biggest issue for for me is that
a) I can't reproduce it in a "regular" python script without fastapi, also not with the use of thread pools etc.
b) I can't reproduce it without joinedload. As soon as I switch to selectinload, there is no memory leak problem whatsoever. 

I'm sorry for my apparent lack of knowledge of the underlying architecture of SQLAlchemy and FastAPI but I'm at a point where it would be super helpful if someone could point me in the right direction. Since this problem only exists with the use of joinedload, I figured that I start with a discussion thread here. 

**Questions:** 
1) Is this a SQLAlchemy or a FastAPI(uvicorn, starlette, etc) problem?
If SQLAlchemy:
 2) Is this considered normal behaviour? 
 - Why does the memory in the worker increase with each call?
 - Why does joinedload use much more memory after the query is done? Are the loaded result lines kept somewhere until the whole result object is "removed"?
 3) Am I doing something very stupid here? (I am in the demo, see the disclaimer below)
 
 
 We use postgresql in production, with psycopg2-binary==2.8.6 as driver. In this example I switched to sqlite. We also use other fastapi versions, but results are quite the same, demo has current ones.
 
 
**Steps to reproduce:**

**ATTENTION!: The query used IS STUPID.** It's only purpose is to maximize the lines loaded via joinedload for demo purposes. I am aware that this returns thousands of result lines from the DB which need to be processed by SQLALchemy and is much worse than using selectinload. I am aware the usage of joinedload is stupid with those relations. However, we have some queries in production where we think that using joined load and returning 10-20 lines is faster/more efficient than using selectinload and having multiple queries for those small result sets. 

Start up the following container and use curl from inside the container:
`while true; do curl -X GET "http://127.0.0.1:8000/companies"; sleep 1; done`

Results with joinedload:
![image](https://github.com/user-attachments/assets/b8951182-0ca4-4e86-a20d-fccbe008ac19)

Results with selectinload:
![image](https://github.com/user-attachments/assets/8fe2f694-3664-4011-aa2a-7a472fb027f7)

I am aware that it might seem that the increase in used memory does stall sometimes or sometimes the memory in use even decreases but overall, it's enough upward tendency that I am concerned and we defenitly have a problem on our prod environment where we use joinedload with much smaller result sets (i.e. after making the tradeoff between having more lines with joinedload vs. more queries with selectinload)

```
Dockerfile:
FROM python:3.10@sha256:aa79a3d35cb9787452dad51e17e4b6e06822a1a601f8b4ac4ddf74f0babcbfd5

RUN apt-get update
RUN apt-get install -y curl

#RUN pip install psycopg2-binary==2.8.6
RUN pip install sqlalchemy==1.4.52
RUN pip install fastapi[all]==0.112.0
RUN pip install gunicorn==22.0.0
RUN pip install psutil

WORKDIR /app

COPY main.py /app/

EXPOSE 8000
ENTRYPOINT ["fastapi", "run", "main.py"]
```

```
import sys
import uuid
from random import choice

import psutil
from fastapi import Depends, FastAPI

# from models import Base
from sqlalchemy import Boolean, Column, DateTime, ForeignKey, Integer, String, create_engine, event, select
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import Session, joinedload, relationship, selectinload, sessionmaker
from sqlalchemy.schema import UniqueConstraint
from sqlalchemy.sql.sqltypes import LargeBinary

DATABASE_URL = "sqlite:///./test1.db"  # or your database URL

process = psutil.Process()

engine = create_engine(DATABASE_URL, connect_args={"check_same_thread": False})
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)


Base = declarative_base()


class User(Base):
    __tablename__ = "user"

    id = Column(String(36), primary_key=True, index=True, nullable=False)
    email = Column(String, nullable=False, unique=True)
    profile_pic = Column(LargeBinary, nullable=True)
    profile_pic_format = Column(String)
    is_active = Column(Boolean, server_default="t", default=True)
    company_id = Column(String, ForeignKey("company.id"))
    company = relationship("Company", back_populates="employees")
    position = Column(String, nullable=True)
    first_name = Column(String, nullable=True)
    last_name = Column(String, nullable=True)

    sync_key = Column(String)


class Company(Base):
    __tablename__ = "company"
    id = Column(String(36), primary_key=True, index=True, nullable=False)
    profile_pic = Column(LargeBinary, nullable=True)
    profile_pic_format = Column(String)
    name = Column(String)
    is_sender = Column(Boolean, server_default="f", default=False, nullable=False)
    is_receiver = Column(Boolean, server_default="f", default=False, nullable=False)
    status_comment = Column(String, nullable=True)
    internal_status_comment = Column(String, nullable=True)
    known_issues = Column(String, nullable=True)
    current_issues = Column(String, nullable=True)
    further_information = Column(String, nullable=True)
    rank = Column(Integer, server_default="0", nullable=False)

    employees = relationship("User", uselist=True, back_populates="company")
    accounting_references = relationship("AccountingReference", uselist=True, back_populates="company")


class AccountingReference(Base):
    __tablename__ = "accounting_reference"
    __table_args__ = (UniqueConstraint("type", "name", "company_id", name="unique_accounting_reference"),)
    id = Column(String(36), primary_key=True, index=True, nullable=False)
    type = Column(String)
    name = Column(String)
    is_active = Column(Boolean, server_default="t", default=True)
    company_id = Column(String, ForeignKey("company.id"))
    company = relationship("Company", back_populates="accounting_references")


Base.metadata.create_all(engine)


def random_string(length=10):
    letters = "abcdefghijklmnopqrstuvwxyz"
    return "".join(choice(letters) for i in range(length))


# Function to create companies with users and accounting references
def create_companies(session, num_companies=30, num_users_per_company=100, num_refs_per_company=100):
    for _ in range(num_companies):
        company = Company(
            id=str(uuid.uuid4()), name=f"Company_{random_string(5)}", profile_pic=b"", profile_pic_format="png"
        )
        session.add(company)

        for _ in range(num_users_per_company):
            user = User(
                id=str(uuid.uuid4()),
                email=f"user_{random_string(10)}@example.com",
                first_name=random_string(5),
                last_name=random_string(5),
                company=company,
                profile_pic=b"",
                profile_pic_format="png",
            )
            session.add(user)

        for _ in range(num_refs_per_company):
            accounting_reference = AccountingReference(
                id=str(uuid.uuid4()), type=random_string(5), name=random_string(10), company=company
            )
            session.add(accounting_reference)


# Create a new session and generate the data
with SessionLocal() as session:
    create_companies(session)
    session.commit()
    session.close()


# print(f"Memory usage 2: {memory_usage_mb:.2f} MB")


def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


app = FastAPI()


def fetch_data(db):
    stmt = (
        select(Company)
        .options(
            joinedload(Company.employees),
            joinedload(Company.accounting_references),
        )
        .distinct()
    )
    results = db.execute(stmt).unique().scalars().all()

    name_l = []
    for comp in results:
        name_l.append(f"{comp.id} - {comp.name} - size {sys.getsizeof(comp)}")

    memory_info = process.memory_info()
    memory_usage_mb = memory_info.rss / (1024**2)
    print(f"Memory usage b4 return: {memory_usage_mb:.2f} MB")
    return name_l


@app.get("/companies")
def read_companies(offset: int = 0, limit: int = 30, db: Session = Depends(get_db)):
    res = fetch_data(db)
    memory_info = process.memory_info()
    # print(f"Some results {str(res)[:20]}...")
    print(f"Memory usage after call: {memory_info.rss / (1024 ** 2):.2f} MB")
    return {"memory": f"{memory_info.rss / (1024 ** 2):.2f} MB"}

```

Much love <3

**Comment by zzzeek** [comment]
thanks for the very complete info, this is the first time I've been given enough to start with that im actually using fastapi.

bit of a mystery so far on this one

**Reply by zzzeek** [reply]
so far this looks like the memory changing is difficult to measure using psutil as there's just natural growth of the Python process going on, but there's no net leak of actual RAM, that is, I'm running top on the process at the same time and while it shows memory doing a real roller coaster, it's not net growing.     

one thing to be extremely wary about is using SQLite in a test of process RAM, because SQlite is itself an in-memory database, and the sqlite3 python driver also does caching stuff internally that has a stair-step memory pattern.  so there is opaque use of memory both in the sqlite embedded library as well as the python driver.   the query here with the JOIN will have a more complex query plan than the selectinload version which may lead to sqlite internally doing more complicated things.

So i've switched this out for mysql instead,  added lengths to all the `String` types and used `true()` / `false()` for the boolean defaults to make it mysql compatible.  that removes all the sqlite embedded stuff from the equation.

I tried removing factors such as query caching, the uniquing of rows (disabling the internal SQLalchemy check), none of that really affected how it behaves so I've ruled those out as factors here.   it looks like the memory jumps only based on fetching the actual objects, so this is object memory overhead.   the size of this object graph doesnt change so there should be no leak in that regard.

you have to look that your "memory b4 return" is a larger number than the "memory after call".  that top "memory b4 return" is not growing at all here.  So this looks more like Python gc is tuning its efforts each time the very large result is lost.

```
Memory usage b4 return: 466.02 MB
Memory usage after call: 197.00 MB
INFO:     127.0.0.1:39016 - "GET /companies HTTP/1.1" 200 OK
Memory usage b4 return: 462.98 MB
Memory usage after call: 196.91 MB
INFO:     127.0.0.1:39024 - "GET /companies HTTP/1.1" 200 OK
Memory usage b4 return: 465.81 MB
Memory usage after call: 195.94 MB
INFO:     127.0.0.1:39040 - "GET /companies HTTP/1.1" 200 OK
Memory usage b4 return: 463.97 MB
Memory usage after call: 198.96 MB
INFO:     127.0.0.1:52058 - "GET /companies HTTP/1.1" 200 OK
Memory usage b4 return: 465.94 MB
Memory usage after call: 199.89 MB
INFO:     127.0.0.1:52064 - "GET /companies HTTP/1.1" 200 OK
Memory usage b4 return: 463.95 MB
Memory usage after call: 201.86 MB
INFO:     127.0.0.1:52068 - "GET /companies HTTP/1.1" 200 OK
Memory usage b4 return: 466.91 MB
Memory usage after call: 207.80 MB
INFO:     127.0.0.1:52076 - "GET /companies HTTP/1.1" 200 OK
Memory usage b4 return: 463.95 MB
Memory usage after call: 204.75 MB
INFO:     127.0.0.1:57398 - "GET /companies HTTP/1.1" 200 OK
Memory usage b4 return: 465.90 MB
Memory usage after call: 204.77 MB
INFO:     127.0.0.1:57414 - "GET /companies HTTP/1.1" 200 OK
Memory usage b4 return: 465.99 MB
Memory usage after call: 207.88 MB
INFO:     127.0.0.1:57426 - "GET /companies HTTP/1.1" 200 OK
Memory usage b4 return: 463.07 MB
Memory usage after call: 208.83 MB
INFO:     127.0.0.1:57430 - "GET /companies HTTP/1.1" 200 OK
Memory usage b4 return: 464.01 MB
Memory usage after call: 202.84 MB
INFO:     127.0.0.1:48896 - "GET /companies HTTP/1.1" 200 OK
Memory usage b4 return: 463.95 MB
Memory usage after call: 202.80 MB
INFO:     127.0.0.1:48898 - "GET /companies HTTP/1.1" 200 OK
Memory usage b4 return: 464.92 MB
Memory usage after call: 206.88 MB
INFO:     127.0.0.1:48902 - "GET /companies HTTP/1.1" 200 OK
Memory usage b4 return: 464.04 MB
Memory usage after call: 207.86 MB
INFO:     127.0.0.1:48916 - "GET /companies HTTP/1.1" 200 OK
Memory usage b4 return: 465.04 MB
Memory usage after call: 211.83 MB
INFO:     127.0.0.1:53156 - "GET /companies HTTP/1.1" 200 OK
Memory usage b4 return: 464.07 MB
Memory usage after call: 210.90 MB
INFO:     127.0.0.1:53166 - "GET /companies HTTP/1.1" 200 OK
Memory usage b4 return: 465.03 MB
Memory usage after call: 212.77 MB
INFO:     127.0.0.1:53172 - "GET /companies HTTP/1.1" 200 OK
Memory usage b4 return: 467.00 MB
Memory usage after call: 212.72 MB
INFO:     127.0.0.1:53184 - "GET /companies HTTP/1.1" 200 OK
Memory usage b4 return: 464.03 MB
Memory usage after call: 210.74 MB
INFO:     127.0.0.1:51402 - "GET /companies HTTP/1.1" 200 OK
Memory usage b4 return: 466.14 MB
Memory usage after call: 210.85 MB
INFO:     127.0.0.1:51414 - "GET /companies HTTP/1.1" 200 OK
Memory usage b4 return: 465.14 MB
Memory usage after call: 210.85 MB
INFO:     127.0.0.1:51424 - "GET /companies HTTP/1.1" 200 OK
Memory usage b4 return: 466.01 MB
Memory usage after call: 212.71 MB
INFO:     127.0.0.1:51432 - "GET /companies HTTP/1.1" 200 OK

```

the bottom "after call" number does slowly creep upwards towards the "b4 return" number, but it's not consistent and it moves upwards more and more slowly.  that's not what a leak usually looks like.  it looks more like Python is tuning its internal GC.    I'll let this run for a few hours more

current script:

```py
import sys
import uuid
from random import choice

import psutil
from fastapi import Depends, FastAPI

import gc
# from models import Base
from sqlalchemy import Boolean, Column, DateTime, ForeignKey, Integer, String, create_engine, event, select, false, true
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import Session, joinedload, relationship, selectinload, sessionmaker
from sqlalchemy.schema import UniqueConstraint
from sqlalchemy.sql.sqltypes import LargeBinary

#DATABASE_URL = "sqlite:///./test1.db"  # or your database URL

DATABASE_URL = "mysql://scott:tiger@localhost/test"

process = psutil.Process()

#engine = create_engine(DATABASE_URL, connect_args={"check_same_thread": False}, query_cache_size=0)
engine = create_engine(DATABASE_URL)
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)


Base = declarative_base()


class User(Base):
    __tablename__ = "user"

    id = Column(String(36), primary_key=True, index=True, nullable=False)
    email = Column(String(50), nullable=False, unique=True)
    profile_pic = Column(LargeBinary, nullable=True)
    profile_pic_format = Column(String(50))
    is_active = Column(Boolean, server_default=true(), default=True)
    company_id = Column(String(50), ForeignKey("company.id"))
    company = relationship("Company", back_populates="employees")
    position = Column(String(50), nullable=True)
    first_name = Column(String(50), nullable=True)
    last_name = Column(String(50), nullable=True)

    sync_key = Column(String(50))


class Company(Base):
    __tablename__ = "company"
    id = Column(String(36), primary_key=True, index=True, nullable=False)
    profile_pic = Column(LargeBinary, nullable=True)
    profile_pic_format = Column(String(50))
    name = Column(String(50))
    is_sender = Column(Boolean, server_default=false(), default=False, nullable=False)
    is_receiver = Column(Boolean, server_default=false(), default=False, nullable=False)
    status_comment = Column(String(50), nullable=True)
    internal_status_comment = Column(String(50), nullable=True)
    known_issues = Column(String(50), nullable=True)
    current_issues = Column(String(50), nullable=True)
    further_information = Column(String(50), nullable=True)
    rank = Column(Integer, server_default="0", nullable=False)

    employees = relationship("User", uselist=True, back_populates="company")
    accounting_references = relationship("AccountingReference", uselist=True, back_populates="company")


class AccountingReference(Base):
    __tablename__ = "accounting_reference"
    __table_args__ = (UniqueConstraint("type", "name", "company_id", name="unique_accounting_reference"),)
    id = Column(String(36), primary_key=True, index=True, nullable=False)
    type = Column(String(50))
    name = Column(String(50))
    is_active = Column(Boolean, server_default=true(), default=True)
    company_id = Column(String(50), ForeignKey("company.id"))
    company = relationship("Company", back_populates="accounting_references")


Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)


def random_string(length=10):
    letters = "abcdefghijklmnopqrstuvwxyz"
    return "".join(choice(letters) for i in range(length))


# Function to create companies with users and accounting references
def create_companies(session, num_companies=30, num_users_per_company=100, num_refs_per_company=100):
    for _ in range(num_companies):
        company = Company(
            id=str(uuid.uuid4()), name=f"Company_{random_string(5)}", profile_pic=b"", profile_pic_format="png"
        )
        session.add(company)

        for _ in range(num_users_per_company):
            user = User(
                id=str(uuid.uuid4()),
                email=f"user_{random_string(10)}@example.com",
                first_name=random_string(5),
                last_name=random_string(5),
                company=company,
                profile_pic=b"",
                profile_pic_format="png",
            )
            session.add(user)

        for _ in range(num_refs_per_company):
            accounting_reference = AccountingReference(
                id=str(uuid.uuid4()), type=random_string(5), name=random_string(10), company=company
            )
            session.add(accounting_reference)


# Create a new session and generate the data
with SessionLocal() as session:
    create_companies(session)
    session.commit()
    session.close()


# print(f"Memory usage 2: {memory_usage_mb:.2f} MB")


def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


app = FastAPI()


def fetch_data(db):
    stmt = (
        select(Company)
        .options(
            joinedload(Company.employees),
            joinedload(Company.accounting_references),
        )
        .distinct()
    )
    results = db.execute(stmt).unique().scalars().all()
    #results = db.execute(stmt).scalars().all()

    name_l = []
    for comp in results:
        name_l.append(f"{comp.id} - {comp.name} - size {sys.getsizeof(comp)}")

    memory_info = process.memory_info()
    memory_usage_mb = memory_info.rss / (1024**2)
    print(f"Memory usage b4 return: {memory_usage_mb:.2f} MB")
    return name_l


@app.get("/companies")
def read_companies(offset: int = 0, limit: int = 30, db: Session = Depends(get_db)):
    res = fetch_data(db)
    del res
    memory_info = process.memory_info()
    # print(f"Some results {str(res)[:20]}...")
    print(f"Memory usage after call: {memory_info.rss / (1024 ** 2):.2f} MB")
    return {"hello": "world"}   # just in case there's some weird page caching by fastapi, shrugs
    #return {"memory": f"{memory_info.rss / (1024 ** 2):.2f} MB"}


```

**Reply by zzzeek** [reply]
top display, the %mem never goes over 2.1%, it goes between about 0.8 and 2.1

![image](https://github.com/user-attachments/assets/c41e728d-f8a9-4046-a5c4-d8f16a97f1df)

**Reply by zzzeek** [reply]
seven minutes later, memory hasnt grown

![image](https://github.com/user-attachments/assets/9b40c5ac-0b96-48e3-b786-adfc3a61f405)

**Reply by makatz** [reply]
Hi zzzeek,

thank you so much for your quick reply. 

First of all, please always assume incompetence on my part. Hence the discussion and not the creation of any issue. 

Some more background info: 

1) We experience this on our prod environment where we can fill ~20-30 GB RAM per day "forever". This problem vanishes when we remove all joinedloads and replace them with selectinloads. Of course those numbers are not super relevant because they are dependend on the number of requests/querys and the size of the loaded datasets.

2) I have a very hard time reproducing this stuff and this was more of a lead and I'm still not sure if SQLAlchemy is the problem.  Between simplification of models, queries, DBs and up/downgrading lib version numbers, it seems like I just can't quite catch it.

3) I have a few assumption why it is difficult for me to display the same results on other environments or in docker containers:
- While we sometimes only load a few rows, they might contain postgres JSONB columns with a lot of data (we're fixing that right now and removing joinedload as well). I have the feeling that the memory which is occupied forever scales with the size of the rows loaded from postgres in our case.
- We have a lot more API workers on there than in any reproducable example. My assumption is that every worker can slowly creep up to the ceiling you are mentioning which obviously makes the problem a lot bigger for us. 
- I'am also assuming that it can also grow larger in one worker when different datasets are loaded (i.e. not always the same companies for example). Although I have no good minimalized example for that. 

4) I tried to use the tracemalloc library to explore what occoupies the memory but I didn't see any changes in memory there. My very dumb assumption here is that this can't find any memory occupied by non-python code (C ?) used in libraries. As menionend, please forgive my lack of understanding of the underlying mechanisms here. 

This is why I was looking for some "guidance" more than I wanted to make SQLAlchemy responsible for leaking memory. You are absolutley right with everything you said. The main questions for me are still:

1) Why the behaviour differs so much when selectinload is used (it doesn't go up in any way, right?)
2) Why is the result "very large"? In my understanding the result object should be exactly the same size (which it is with the naive methods of testing that, i.e. os.getsizeof) with both loading strategies?  I get that there are much more rows loaded with joinedload but after "parsing them" into the SQLALchemy objects/result object, this memory should be free again, right?

**Reply by zzzeek** [reply]
> Hi zzzeek,
> 
> thank you so much for your quick reply.
> 
> First of all, please always assume incompetence on my part. Hence the discussion and not the creation of any issue.
> 
> Some more background info:
> 
>     1. We experience this on our prod environment where we can fill ~20-30 GB RAM per day "forever". 

it's important that you clarify what this means.  is it that the process goes to 30G of RAM and stays there, and never shrinks?  that's normal for Python when you have a very large set of memory allocated over and over again.  this is in no way referred to as a "leak"

if OTOH the process **continues to grow, unbounded, until the server crashes**, that's a memory leak.  

these are two really different scenarios, so if we can identify which one, I would produce more guidance for that specific scenario.



> This problem vanishes when we remove all joinedloads and replace them with selectinloads. 

there's a simple reason for this which is that the JOIN query you have will produce an enormous number of rows that need to be "uniqued" for the joined eager load.   joined eager load the ORM has to fetch and process 300K rows:

```
sqlite> select count(*) from ( SELECT anon_1.id, anon_1.profile_pic, anon_1.profile_pic_format, anon_1.name, anon_1.is_sender, anon_1.is_receiver, anon_1.status_comment, anon_1.internal_status_comment, anon_1.known_issues, anon_1.current_issues, anon_1.further_information, anon_1.rank, user_1.id AS id_1, user_1.email, user_1.profile_pic AS profile_pic_1, user_1.profile_pic_format AS profile_pic_format_1, user_1.is_active, user_1.company_id, user_1.position, user_1.first_name, user_1.last_name, user_1.sync_key, accounting_reference_1.id AS id_2, accounting_reference_1.type, accounting_reference_1.name AS name_1, accounting_reference_1.is_active AS is_active_1, accounting_reference_1.company_id AS company_id_1 
FROM (SELECT DISTINCT company.id AS id, company.profile_pic AS profile_pic, company.profile_pic_format AS profile_pic_format, company.name AS name, company.is_sender AS is_sender, company.is_receiver AS is_receiver, company.status_comment AS status_comment, company.internal_status_comment AS internal_status_comment, company.known_issues AS known_issues, company.current_issues AS current_issues, company.further_information AS further_information, company.rank AS rank 
FROM company) AS anon_1 LEFT OUTER JOIN user AS user_1 ON anon_1.id = user_1.company_id LEFT OUTER JOIN accounting_reference AS accounting_reference_1 ON anon_1.id = accounting_reference_1.company_id
);
300000

```

you can see this query using echo=True

with selectinload it only needs to process the exact number of rows as there are objects , 6030 rows:

```
sqlite> select count(*) from user;
3000
sqlite> select count(*) from company;
30
sqlite> select count(*) from accounting_reference;
3000

```

**Comment by CaselIT** [comment]
@makatz a few things to try
Can you use a more recent version of psycog2? The ine you pinned is very old. Maybe trying also psycog3 could be an option 

Can you try using fastapi with an async def function? When you use a sync function it tries to do very clever things that IIRC caused memory leaks in older versions of the library (not sufe if it was in combination with sqlalchemy or not)
You will need to change the driver you use, psycopg3 or asyncpg should be equally good
